Update with army and defenders

mishoto opened the pull request:
Tests for Army, WarriorFactory and Battle are in progress!

mishoto committed:
make attack constant
refactor hitByKnight() to hit() for working with Knight & Warrior
add private setHealth()

mishoto committed:
make attack constant
remove health field, constructor, getHealth() and hitByWarrior()

mishoto committed:
remove fields and constructor
add check for health in while loop

mishoto committed:
add test dependencies

mishoto committed:
minor refactoring

mishoto committed:
add @tests for Duel, Knight & Warrior

mishoto committed:
create Army

mishoto committed:
create Actions interface

mishoto committed:
create Battle

mishoto committed:
create Defender

mishoto committed:
refactor Duel to implement Actions

mishoto committed:
introduce instance of Duel for testing fight

mishoto committed:
create Factory for characters

mishoto committed:
move army to models package and refactor with builder for army creation with character factory

mishoto committed:
moved to renamed models package
rename characters to models package

mishoto committed:
create concrete WarriorFactory

mishoto committed:
implement Cloneable and extend Character

mishoto committed:
moved to renamed package

mishoto committed:
moved to renamed package

mishoto committed:
delete Actions

mishoto committed:
remove parameter from isAlive method

mishoto committed:
create FightStrategy interface

mishoto committed:
refactor fight with new static method

mishoto committed:
refactor isAlive -> remove health as parameter

mishoto committed:
add getTroopIterator(),
getArmy() to flatten nested Lists
getNextWarrior()

mishoto committed:
add logic to fight()

mishoto committed:
add constructor with setter

mishoto committed:
refactor hit() for the defender condition

mishoto committed:
create Tests for Defender

KaloyanTsvetkov commented on pom.xml:
```diff
@@ -12,5 +12,21 @@
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <version>4.13.2</version>
```
Why you are using junit 4 dependency?
As I know you should use junit5 for the unit tests?

KaloyanTsvetkov commented on src/main/java/edu/mihail/factories/WarriorFactory.java:
```diff
@@ -0,0 +1,37 @@
+            case DEFENDER:
+                return new Defender();
+            default:
+                throw new IllegalArgumentException("There is no such warrior type!");
```
Please create your own Exception.

KaloyanTsvetkov commented on src/main/java/edu/mihail/models/Army.java:
```diff
@@ -0,0 +1,56 @@
+import java.util.stream.Collectors;
+
+
+public class Army {
```
I can not find your unit tests for this class.

KaloyanTsvetkov commented on src/main/java/edu/mihail/fightscene/Battle.java:
```diff
@@ -0,0 +1,32 @@
+
+import java.util.List;
+
+public class Battle implements FightStrategy<Army> {
```
I can not find your unit tests for this class.

KaloyanTsvetkov commented on src/main/java/edu/mihail/models/Army.java:
```diff
@@ -0,0 +1,56 @@
+    private final Iterator<List<Warrior>> troopIterator;
+
+    public static class Builder {
+        private final List<List<Warrior>> troop = new LinkedList<>();
```
Still can not understand why you need to be List<List<Warrior>>.
why not only List<Warrior>?

KaloyanTsvetkov commented on src/test/java/edu/mihail/models/DefenderTest.java:
```diff
@@ -0,0 +1,26 @@
+        warriorOne.hit(defenderOne);
+        Assertions.assertEquals(57, defenderOne.getHealth());
+        Assertions.assertEquals(50, warriorOne.getHealth());
+
```
warriorOne.hit & defenderOne.hit should be two separate tests.
Please use setUp method: https://stackoverflow.com/questions/512184/best-practice-initialize-junit-class-fields-in-setup-or-at-declaration

KaloyanTsvetkov commented on src/test/java/edu/mihail/models/KnightTest.java:
```diff
@@ -0,0 +1,21 @@
+
+class KnightTest {
+
+    Warrior knight = new Knight();
```
Please use setUp method.

KaloyanTsvetkov commented on src/test/java/edu/mihail/models/WarriorTest.java:
```diff
@@ -0,0 +1,87 @@
+
+class WarriorTest {
+
+    private final Warrior warriorOne = new Warrior();
```
Please use setUp method

mishoto committed:
Merge branch 'main' into update-with-army-and-defenders

# Conflicts:
#	src/main/java/edu/mihail/characters/Knight.java
#	src/main/java/edu/mihail/characters/Warrior.java
#	src/main/java/edu/mihail/fightscene/Duel.java
#	src/test/java/edu/mihail/fightscene/DuelTest.java

mishoto committed:
minor changes

mishoto committed:
create a mock inner class for testing isAlive()

mishoto committed:
delete duplicates from main

mishoto committed:
remove junit 4.13.2

mishoto committed:
use warrior.getHealth() in hit()

mishoto committed:
remove getHealth()

mishoto committed:
create additional test cases

mishoto committed:
add vampire and lancer

mishoto committed:
minor refactoring

mishoto committed:
minor refactoring

mishoto committed:
create Lancer & Vampire Tests

mishoto committed:
add lancer & vampire types

mishoto committed:
create lancer & vampire characters

mishoto committed:
refactor hit()
add defence field
add logic for: calculateWarriorHealth()
calculateWarriorDamage()

mishoto committed:
minor changes
